api/prometheus/v1/remote/remote_api.go

```diff
@@ -0,0 +1,376 @@
+		r.reqBuf = r.reqBuf[:0]
+		r.reqBuf, err = (proto.MarshalOptions{}).MarshalAppend(r.reqBuf, msg)
+		if err != nil {
+			return WriteResponseStats{}, fmt.Errorf("encoding request %w", err)
+		}
+	}
+
+	payload, err := compressPayload(&r.comprBuf, r.opts.compression, r.reqBuf)
+	if err != nil {
+		return WriteResponseStats{}, fmt.Errorf("compressing %w", err)
+	}
+
+	// Since we retry writes we need to track the total amount of accepted data
+	// across the various attempts.
+	accumulatedStats := WriteResponseStats{}
+
+	b := backoff.New(ctx, r.opts.backoff)
+	for {
+		rs, err := r.attemptWrite(ctx, r.opts.compression, cType, payload, b.NumRetries())
+		accumulatedStats = accumulatedStats.Add(rs)
+		if err == nil {
+			// Check the case mentioned in PRW 2.0.
+			// https://prometheus.io/docs/specs/remote_write_spec_2_0/#required-written-response-headers.
+			if cType == WriteProtoFullNameV2 && !accumulatedStats.Confirmed && accumulatedStats.NoDataWritten() {
+				// TODO(bwplotka): Allow users to disable this check or provide their stats for us to know if it's empty.
+				return accumulatedStats, fmt.Errorf("sent v2 request; "+
+					"got 2xx, but PRW 2.0 response header statistics indicate %v samples, %v histograms "+
+					"and %v exemplars were accepted; assumining failure e.g. the target only supports "+
+					"PRW 1.0 prometheus.WriteRequest, but does not check the Content-Type header correctly",
+					accumulatedStats.Samples, accumulatedStats.Histograms, accumulatedStats.Exemplars,
+				)
+			}
+			// Success!
+			// TODO(bwplotka): Debug log with retry summary?
+			return accumulatedStats, nil
+		}
+
+		var retryableErr retryableError
+		if !errors.As(err, &retryableErr) {
+			// TODO(bwplotka): More context in the error e.g. about retries.
+			return accumulatedStats, err
+		}
+
+		if !b.Ongoing() {
+			// TODO(bwplotka): More context in the error e.g. about retries.
+			return accumulatedStats, err
+		}
+
+		backoffDelay := b.NextDelay() + retryableErr.RetryAfter()
+		r.opts.logger.Error("failed to send remote write request; retrying after backoff", "err", err, "backoff", backoffDelay)
+		select {
+		case <-ctx.Done():
+			// TODO(bwplotka): More context in the error e.g. about retries.
+			return WriteResponseStats{}, ctx.Err()
+		case <-time.After(backoffDelay):
+			// Retry.
+		}
+	}
+}
+
+func compressPayload(tmpbuf *[]byte, enc Compression, inp []byte) (compressed []byte, _ error) {
+	switch enc {
+	case SnappyBlockCompression:
+		compressed = snappy.Encode(*tmpbuf, inp)
+		if n := snappy.MaxEncodedLen(len(inp)); n > len(*tmpbuf) {
+			// grow the buffer for the next time.
+			*tmpbuf = make([]byte, n)
+		}
+		return compressed, nil
+	default:
+		return compressed, fmt.Errorf("unknown compression scheme [%v]", enc)
+	}
+}
+
+func (r *API) attemptWrite(ctx context.Context, compr Compression, proto WriteProtoFullName, payload []byte, attempt int) (WriteResponseStats, error) {
+	u := r.client.URL("api/v1/write", nil)
```

Comment:
this should be configurable imo, not all remote write receivers use the same endpoint